### PR TITLE
BUG: silently ignoring dtype kwarg in Index.__new__

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #xxxx
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

ATM we handle object but not others, xref #21311

### CODE REVIEW

**Comment by jreback** [comment]
nice cleanup. needs a rebase. should we consider deprecating `dtype=` keyword in the constructor?

**Comment by jbrockmendel** [comment]
> should we consider deprecating dtype= keyword in the constructor?

my hope/goal is for `Index.__new__`, `Series.__init__`, and pd.array` to share most of their code, so probably not

File: pandas/tests/indexes/test_index_new.py [hunk 1]
```diff
@@ -122,6 +126,80 @@ def test_constructor_mixed_nat_objs_infers_object(self, swap_objs):
         tm.assert_index_equal(Index(np.array(data, dtype=object)), expected)
 
 
+class TestDtypeEnforced:
```
Comment (jreback) [inline pandas/tests/indexes/test_index_new.py:129]: slightly OT: wha tis the purpose of this specific test file, should this not be in base_class/ or maybe test_construction.py?

Comment (jbrockmendel) [reply pandas/tests/indexes/test_index_new.py:129]: this file is aimed at `Index.__new__`, in particular checking it returns the expected subclass

Comment (jbrockmendel) [reply pandas/tests/indexes/test_index_new.py:129]: is the file docstring unclear?

Comment (jreback) [reply pandas/tests/indexes/test_index_new.py:129]: yes, but the *name* of the file is not super clear as 'new' indicates not ```___new__```, would probably rename to test_construction.py

Comment (jbrockmendel) [reply pandas/tests/indexes/test_index_new.py:129]: ok, ill look at renaming in upcoming tests pass

**Pull request merged**